libs/dkg/src/rbc.cpp

```diff
@@ -0,0 +1,571 @@
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "core/logging.hpp"
+#include "crypto/hash.hpp"
+#include "crypto/sha256.hpp"
+#include "dkg/dkg_service.hpp"
+#include "dkg/rbc.hpp"
+
+namespace fetch {
+namespace dkg {
+namespace rbc {
+
+constexpr char const *LOGGING_NAME = "RBC";
+
+/**
+ * Helper function to compute truncation of message hash to 8 bytes. Truncation from left side.
+ */
+TruncatedHash MessageHash(const SerialisedMessage &msg)
+{
+  byte_array::ByteArray msg_hash_256{crypto::Hash<crypto::SHA256>(msg)};
+  return msg_hash_256.SubArray(24);
+}
+
+std::string RBC::MsgTypeToString(MsgType msg_type)
+{
+  switch (msg_type)
+  {
+  case MsgType::R_SEND:
+    return "r_send";
+  case MsgType::R_ECHO:
+    return "r_echo";
+  case MsgType::R_READY:
+    return "r_ready";
+  case MsgType::R_REQUEST:
+    return "r_request";
+  case MsgType::R_ANSWER:
+    return "r_answer";
+  default:
+    assert(false);
+    return "";
+  }
+}
+
+/**
+ * Creates instance of RBC
+ *
+ * @param endpoint The muddle endpoint to communicate on
+ * @param address The muddle endpoint address
+ * @param cabinet Set of muddle addresses that form RBC network
+ * @param threshold Threshold number of Byzantine peers
+ * @param dkg
+ */
+RBC::RBC(Endpoint &endpoint, MuddleAddress address, CabinetMembers const &cabinet,
+         std::size_t &threshold, DkgService &dkg_service)
+  : address_{std::move(address)}
+  , endpoint_{endpoint}
+  , current_cabinet_{cabinet}
+  , threshold_{threshold}
+  , dkg_service_{dkg_service}
+  , rbc_subscription_(endpoint.Subscribe(SERVICE_DKG, CHANNEL_BROADCAST))
+{
+  // Set subscription for rbc
+  rbc_subscription_->SetMessageHandler([this](MuddleAddress const &from, uint16_t, uint16_t,
+                                              uint16_t, muddle::Packet::Payload const &payload,
+                                              MuddleAddress) {
+    RBCSerializer serialiser(payload);
+
+    // Deserialize the RBCEnvelop
+    RBCEnvelop env;
+    serialiser >> env;
+
+    // Dispatch the event
+    OnRBC(from, env);
+  });
+}
+
+/**
+ * Resets the RBC for a new cabinet
+ */
+void RBC::ResetCabinet()
+{
+  assert(!current_cabinet_.empty());
+  assert(current_cabinet_.find(address_) != current_cabinet_.end());
+  assert(current_cabinet_.size() > 3 * threshold_);
+  std::lock_guard<std::mutex> lock{mutex_broadcast_};
+  id_ = static_cast<uint32_t>(
```

Comment:
can id_ ever change ? current_cab is const

Reply:
Yes id_ can change if the cabinet is reset inside DkgService. current_cabinet should be a const reference in RBC.

Reply:
Oh I see - it looks like id_ is uninitialized until reset is called at which point it will be set to a value that can never change (since current_cabinet can't change) - would it make sense to set id_ in the constructor instead?

Reply:
Ignore me - I see now that the current cabinet could be changed in the DKG service and the RBC just holds a reference to that, so it can change (I misread it as just const and not const &). Note this could be the source of race conditions depending on when the cabinet gets reset.